File: 백준/Silver/10973. 이전 순열/이전 순열.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int T = Integer.parseInt(reader.readLine());
        int[] arr = new int[T];
        int idx =0;
        for(String num: reader.readLine().split(" ")) {
            arr[idx++] = Integer.parseInt(num);
        }

        boolean hasPrev = prevPermutation(arr, T);

        if(hasPrev) {
            for(int num: arr) {
                writer.write(num + " ");
            }
        }else{
            writer.write(-1 +"");
        }

        writer.flush();
        writer.close();
    }

    public static boolean prevPermutation(int[] arr, int N) {
        int i = N-1;
        while(i > 0 && arr[i-1] < arr[i]) i-=1;
        if(i == 0) return false;
        int j = N-1;
        while(arr[i-1] < arr[j] && arr[j-1] < arr[j]) j-=1;
        swap(arr, i-1, j);
        j = N-1;
        while(i < j) {
            swap(arr, i, j);
            i +=1;
            j -=1;
        }

        return true;
    }

    public static void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: 백준/Silver/10973. 이전 순열/이전 순열.py
import sys

input = sys.stdin.readline


def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]


def next_permutation(arr, n):
    i = n - 1
    while i > 0 and arr[i - 1] < arr[i]:
        i -= 1
    if i == 0:
        return False

    j = n - 1
    while arr[i - 1] < arr[j]:
        j -= 1
    swap(arr, i-1, j)

    j = n - 1
    while i < j:
        swap(arr, i, j)
        j -= 1
        i += 1

    return True


def main():
    n = int(input())
    arr = list(map(int, input().split()))
    result = next_permutation(arr, n)

    if result:
        for num in arr:
            print(num, end=" ")
    else:
        print(-1)


main()
